README.rst
.. image:: https://requires.io/github/Dennis-van-Gils/project-Dodecahedron/requirements.svg?branch=main
    :target: https://requires.io/github/Dennis-van-Gils/project-Dodecahedron/requirements/?branch=main
    :alt: Requirements Status
.. image:: https://img.shields.io/badge/code%20style-black-000000.svg
    :target: https://github.com/psf/black
.. image:: https://img.shields.io/badge/License-MIT-purple.svg
    :target: https://github.com/Dennis-van-Gils/project-Dodecahedron/blob/master/LICENSE.txt

Twente Dodecahedron
===================
*A Physics of Fluids project.*

Control program for the Twente Dodecahedron: a scientific setup in order to
study isotropic droplet-laden turbulence. This program logs the temperature
and humidity readings and it controls the circulating bath that is connected to
the setup's heat exhanging plates.

- Github: https://github.com/Dennis-van-Gils/project-Dodecahedron

.. image:: https://raw.githubusercontent.com/Dennis-van-Gils/project-Dodecahedron/master/images/screenshot.png

Hardware
========
* Adafruit #3857: Adafruit Feather M4 Express - Featuring ATSAMD51 Cortex M4
* Adafruit #381: Waterproof 1-Wire DS18B20 Compatible Digital temperature sensor
* Pimoroni PIM472: BME280 Breakout - Temperature, Pressure, Humidity Sensor
* Julabo FP51-SL: Ultra-Low Refrigerated / Heating Circulator

Instructions
============
Download the `latest release <https://github.com/Dennis-van-Gils/project-Dodecahedron/releases/latest>`_
and unpack to a folder onto your drive.

Flashing the firmware
---------------------

Double click the reset button of the Feather while plugged into your PC. This
will mount a drive called `FEATHERBOOT`. Copy
`src_mcu/_build_Feather_M4/CURRENT.UF2 <https://github.com/Dennis-van-Gils/project-Dodecahedron/raw/main/src_mcu/_build_Feather_M4/CURRENT.UF2>`_
onto the Featherboot drive. It will restart automatically with the new
firmware.

Running the application
-----------------------

Preferred Python distributions:
    * `Anaconda <https://www.anaconda.com>`_
    * `Miniconda <https://docs.conda.io/en/latest/miniconda.html>`_

Open `Anaconda Prompt` and navigate to the unpacked folder. Run the following to
install the necessary packages: ::

    cd src_python
    pip install -r requirements.txt
    
Now you can run the application: ::

    python main.py

LED status lights
=================

* Solid blue: Booting and setting up
* Solid green: Ready for communication
* Flashing green: Sensor data is being send over USB


CHANGELOG.rst
Changelog
=========

1.0.0 (2021-03-02)
------------------
* Initial release


src_python/dodeca_check.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""Scan for all log files acquired by the Twente Dodecahedron control program
in the current folder. Those that are missing a plot figure will be processed.

Useful tool for quick inspection.
"""
__author__ = "Dennis van Gils"
__authoremail__ = "vangils.dennis@gmail.com"
__url__ = "https://github.com/Dennis-van-Gils/project-Dodecahedron"
__date__ = "03-03-2021"
__version__ = "1.0"

import os
import re

from dodeca_read_log import read_log
from dodeca_plot_log import plot_log

# ------------------------------------------------------------------------------
#   Main
# ------------------------------------------------------------------------------

if __name__ == "__main__":
    my_path = os.getcwd()
    file_list = [
        f
        for f in os.listdir(my_path)
        if os.path.isfile(os.path.join(my_path, f))
    ]

    for filename in file_list:
        # Look for files matching: ######_###### [+any extra chars] .txt
        p = re.compile("\d{6}_\d{6}(.*?)\.(txt|TXT)$")
        if p.match(filename):
            # Found a matching file
            # Now check if the same filename exists ending with .png
            filename_png = filename[0:-4] + ".png"

            if not os.path.isfile(filename_png):
                # Figure does not yet exists. Create.
                print("Reading file: %s" % filename)
                log = read_log(filename)
                plot_log(log)


src_python/dodeca_plot_log.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""Plot timeseries of a log file acquired by the Twente Dodecahedron control
program. The plot will be saved as image to disk.
"""
__author__ = "Dennis van Gils"
__authoremail__ = "vangils.dennis@gmail.com"
__url__ = "https://github.com/Dennis-van-Gils/project-Dodecahedron"
__date__ = "03-03-2021"
__version__ = "1.0"

import sys
import os
import numpy as np

import tkinter
from tkinter import filedialog

import matplotlib as mpl
import matplotlib.pyplot as plt

from dodeca_read_log import read_log, Log

# Characters
CHAR_PM = u"\u00B1"
CHAR_DEG = u"\u00B0"

# Colors
cm = (
    np.array(
        [
            [255, 255, 0],
            [252, 15, 192],
            [0, 255, 255],
            [255, 255, 255],
            [255, 127, 39],
            [0, 255, 0],
        ]
    )
    / 255
)

# ------------------------------------------------------------------------------
#   plot_log
# ------------------------------------------------------------------------------


def plot_log(log: Log):
    """
    Args:
        log (dodeca_read_log.Log): Log data structure
    """

    # --------------------------------------------------------------------------
    #   Prepare figure
    # --------------------------------------------------------------------------

    mpl.style.use("dark_background")
    mpl.rcParams["font.size"] = 12
    # mpl.rcParams['font.weight'] = "bold"
    mpl.rcParams["axes.titlesize"] = 14
    mpl.rcParams["axes.labelsize"] = 14
    mpl.rcParams["axes.titleweight"] = "bold"
    mpl.rcParams["axes.formatter.useoffset"] = False
    # mpl.rcParams["axes.labelweight"] = "bold"
    mpl.rcParams["lines.linewidth"] = 2
    mpl.rcParams["grid.color"] = "0.25"

    fig1 = plt.figure(figsize=(16, 10), dpi=90)
    fig1.canvas.set_window_title("%s" % log.filename)

    ax1 = fig1.add_subplot(4, 1, 1)
    ax2 = fig1.add_subplot(4, 1, 2, sharex=ax1)
    ax3 = fig1.add_subplot(4, 1, 3, sharex=ax1)
    ax4 = fig1.add_subplot(4, 1, 4, sharex=ax1)

    # Julabo temperatures
    ax1.plot(
        log.time, log.Julabo_setp, "-", color=cm[4], label=("Julabo setp."),
    )
    ax1.plot(
        log.time, log.Julabo_bath, "-", color=cm[5], label=("Julabo bath"),
    )

    ax1.set_title("%s\nJulabo temperatures" % (log.filename))
    ax1.set_xlabel("time (s)")
    ax1.set_ylabel("temperature (%sC)" % CHAR_DEG)
    ax1.grid(True)

    # Arduino temperatures
    ax2.plot(log.time, log.DS_temp, color=cm[0], label="DS temp.")
    ax2.plot(log.time, log.BME_temp, color=cm[1], label="BME temp.")

    ax2.set_title("Arduino temperatures (%s 0.5 K)" % CHAR_PM)
    ax2.set_xlabel("time (s)")
    ax2.set_ylabel("temperature (%sC)" % CHAR_DEG)
    ax2.grid(True)

    # Arduino humitidy
    ax3.plot(log.time, log.BME_humi, color=cm[2], label="BME humi.")

    ax3.set_title("Humidity (%s 3 %%)" % CHAR_PM)
    ax3.set_xlabel("time (s)")
    ax3.set_ylabel("humidity (%)")
    ax3.grid(True)

    # Arduino pressure
    ax4.plot(log.time, log.BME_pres, color=cm[3], label="BME pres.")

    ax4.set_title("Pressure (%s 1 mbar)" % CHAR_PM)
    ax4.set_xlabel("time (s)")
    ax4.set_ylabel("pressure (mbar)")
    ax4.grid(True)

    # --------------------------------------------------------------------------
    #   Final make-up
    # --------------------------------------------------------------------------

    ax_w = 0.9
    ax_h = 0.15

    ax1.set_position([0.08, 0.79, ax_w, ax_h])
    ax2.set_position([0.08, 0.55, ax_w, ax_h])
    ax3.set_position([0.08, 0.31, ax_w, ax_h])
    ax4.set_position([0.08, 0.07, ax_w, ax_h])

    ax1.legend(loc="upper left")
    ax2.legend(loc="upper left")

    # --------------------------------------------------------------------------
    #   Save figure
    # --------------------------------------------------------------------------

    img_format = "png"
    fn_save = "%s.%s" % (log.filename, img_format)
    plt.savefig(
        fn_save,
        dpi=90,
        orientation="portrait",
        format=img_format,
        transparent=False,
    )
    print("Saved image: %s" % fn_save)


# ------------------------------------------------------------------------------
#   Main
# ------------------------------------------------------------------------------

if __name__ == "__main__":
    # Check for optional input arguments
    filename_supplied = False
    for arg in sys.argv[1:]:
        filename = arg
        filename_supplied = True

    root = tkinter.Tk()
    root.withdraw()  # Disable root window

    if not filename_supplied:
        filename = filedialog.askopenfilename(
            initialdir=os.getcwd(),
            title="Select data file",
            filetypes=(("text files", "*.txt"), ("all files", "*.*")),
        )
    if filename == "":
        sys.exit(0)

    root.destroy()  # Close file dialog

    print("Reading file: %s" % filename)
    plot_log(read_log(filename))
    plt.show()


src_python/dodeca_read_log.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""Reads in a log file acquired with the Twente Dodecahedron control program.
A 2nd order Butterworth low-pass filter with a cut-off frequency of 0.1 Hz and
zero-phase distortion will be applied to the DS18B20 and BME280 timeseries.
Validated.
"""
__author__ = "Dennis van Gils"
__authoremail__ = "vangils.dennis@gmail.com"
__url__ = "https://github.com/Dennis-van-Gils/project-Dodecahedron"
__date__ = "03-03-2021"
__version__ = "1.0"

import numpy as np
from scipy import signal
from pathlib import Path


class Log:
    def __init__(self):
        self.filename = ""
        self.header = [""]
        self.time = np.array([])
        self.DS_temp = np.array([])
        self.BME_temp = np.array([])
        self.BME_humi = np.array([])
        self.BME_pres = np.array([])
        self.Julabo_setp = np.array([])
        self.Julabo_bath = np.array([])


def read_log(filepath=None, apply_lowpass_filter: bool = True):
    """Reads in a log file acquired with the Twente Dodecahedron control
    program.

    Args:
        filepath (pathlib.Path, str):
            Path to the data file to open.

        apply_lowpass_filter (bool, default=True):
            Apply a 2nd order Butterworth low-pass filter with a cut-off
            frequency of 0.1 Hz and zero-phase distortion to the DS18B20 and
            BME280 timeseries?

    Returns: instance of Log class
    """
    if isinstance(filepath, str):
        filepath = Path(filepath)

    if not isinstance(filepath, Path):
        raise Exception(
            "Wrong type passed to read_log(). "
            "Should be (str) or (pathlib.Path)."
        )

    if not filepath.is_file():
        raise Exception("File can not be found\n %s" % filepath.name)

    with filepath.open() as f:
        log = Log()

        # Scan the first lines for the start of the header and data sections
        MAX_LINES = 100  # Stop scanning after this number of lines
        str_header = []
        success = False
        for i_line in range(MAX_LINES):
            str_line = f.readline().strip()

            if str_line.upper() == "[HEADER]":
                # Simply skip
                pass
            elif str_line.upper() == "[DATA]":
                # Found data section. Exit loop.
                i_line_data = i_line
                success = True
                break
            else:
                # We must be in the header section now
                str_header.append(str_line)

        if not success:
            raise Exception(
                "Incorrect file format. Could not find [DATA] " "section."
            )

        # Read in all data columns including column names
        tmp_table = np.genfromtxt(
            filepath.name,
            delimiter="\t",
            names=True,
            skip_header=i_line_data + 2,
        )

        # Rebuild into a Matlab style 'struct'
        log.filename = filepath.name[0:-4]
        log.header = str_header
        log.time = tmp_table["time"]
        log.DS_temp = tmp_table["DS_temp"]
        log.BME_temp = tmp_table["BME_temp"]
        log.BME_humi = tmp_table["BME_humi"]
        log.BME_pres = tmp_table["BME_pres"]
        log.Julabo_setp = tmp_table["Julabo_setp"]
        log.Julabo_bath = tmp_table["Julabo_bath"]

        if apply_lowpass_filter:
            # Apply low-pass filtering to specific timeseries
            f_s = 1 / np.mean(
                np.diff(log.time)
            )  # Original sampling frequency [Hz]
            f3dB_LP = 0.1  # Low-pass cut-off frequency: 0.1 [Hz]
            filt_b, filt_a = signal.butter(2, f3dB_LP / (f_s / 2), "lowpass")

            # Fill in the occasional NaN's in the DS_temp signal
            mask = np.isnan(log.DS_temp)
            log.DS_temp[mask] = np.interp(
                np.flatnonzero(mask), np.flatnonzero(~mask), log.DS_temp[~mask]
            )

            log.DS_temp = signal.filtfilt(filt_b, filt_a, log.DS_temp)
            log.BME_temp = signal.filtfilt(filt_b, filt_a, log.BME_temp)
            log.BME_humi = signal.filtfilt(filt_b, filt_a, log.BME_humi)
            log.BME_pres = signal.filtfilt(filt_b, filt_a, log.BME_pres)
            # log.Julabo_setp = signal.filtfilt(filt_b, filt_a, log.Julabo_setp)
            # log.Julabo_bath = signal.filtfilt(filt_b, filt_a, log.Julabo_bath)

    return log


src_python/main.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""Twente Dodecahedron
"""
__author__ = "Dennis van Gils"
__authoremail__ = "vangils.dennis@gmail.com"
__url__ = "https://github.com/Dennis-van-Gils/project-Dodecahedron"
__date__ = "03-03-2021"
__version__ = "1.0"
# pylint: disable=bare-except, broad-except

import os
import sys
import time

import numpy as np
import psutil

from PyQt5 import QtCore, QtGui
from PyQt5 import QtWidgets as QtWid
from PyQt5.QtCore import QDateTime
import pyqtgraph as pg

from dvg_debug_functions import tprint, dprint, print_fancy_traceback as pft
from dvg_pyqt_controls import (
    create_Toggle_button,
    SS_TEXTBOX_READ_ONLY,
    SS_GROUP,
)
from dvg_pyqt_filelogger import FileLogger
from dvg_pyqtgraph_threadsafe import (
    HistoryChartCurve,
    LegendSelect,
    PlotManager,
)

from dvg_devices.Arduino_protocol_serial import Arduino
from dvg_devices.Julabo_circulator_protocol_RS232 import Julabo_circulator
from dvg_devices.Julabo_circulator_qdev import Julabo_circulator_qdev
from dvg_qdeviceio import QDeviceIO


TRY_USING_OPENGL = True
if TRY_USING_OPENGL:
    try:
        import OpenGL.GL as gl  # pylint: disable=unused-import
    except:
        print("OpenGL acceleration: Disabled")
        print("To install: `conda install pyopengl` or `pip install pyopengl`")
    else:
        print("OpenGL acceleration: Enabled")
        pg.setConfigOptions(useOpenGL=True)
        pg.setConfigOptions(antialias=True)
        pg.setConfigOptions(enableExperimental=True)

# Global pyqtgraph configuration
# pg.setConfigOptions(leftButtonPan=False)
pg.setConfigOption("foreground", "#EEE")

# Constants
# fmt: off
DAQ_INTERVAL_MS    = 1000  # [ms]
CHART_INTERVAL_MS  = 500   # [ms]
CHART_HISTORY_TIME = 7200  # [s]
# fmt: on

# Show debug info in terminal? Warning: Slow! Do not leave on unintentionally.
DEBUG = False


def get_current_date_time():
    cur_date_time = QDateTime.currentDateTime()
    return (
        cur_date_time.toString("dd-MM-yyyy"),  # Date
        cur_date_time.toString("HH:mm:ss"),  # Time
        cur_date_time.toString("yyMMdd_HHmmss"),  # Reverse notation date-time
    )


# ------------------------------------------------------------------------------
#   Arduino state
# ------------------------------------------------------------------------------


class State(object):
    """Reflects the actual readings, parsed into separate variables, of the
    Arduino. There should only be one instance of the State class.
    """

    def __init__(self):
        self.time = np.nan  # [s]
        self.ds_temp = np.nan  # ['C]
        self.bme_temp = np.nan  # ['C]
        self.bme_humi = np.nan  # [%]
        self.bme_pres = np.nan  # [bar]


state = State()

# ------------------------------------------------------------------------------
#   MainWindow
# ------------------------------------------------------------------------------


class MainWindow(QtWid.QWidget):
    def __init__(self, parent=None, **kwargs):
        super().__init__(parent, **kwargs)

        self.setWindowTitle("Dodecahedron control")
        self.setGeometry(350, 60, 1200, 900)
        self.setStyleSheet(SS_TEXTBOX_READ_ONLY + SS_GROUP)

        # -------------------------
        #   Top frame
        # -------------------------

        # Left box
        self.qlbl_update_counter = QtWid.QLabel("0")
        self.qlbl_DAQ_rate = QtWid.QLabel("DAQ: nan Hz")
        self.qlbl_DAQ_rate.setStyleSheet("QLabel {min-width: 7em}")

        vbox_left = QtWid.QVBoxLayout()
        vbox_left.addWidget(self.qlbl_update_counter, stretch=0)
        vbox_left.addStretch(1)
        vbox_left.addWidget(self.qlbl_DAQ_rate, stretch=0)

        # Middle box
        self.qlbl_title = QtWid.QLabel(
            "Dodecahedron control",
            font=QtGui.QFont("Palatino", 14, weight=QtGui.QFont.Bold),
        )
        self.qlbl_title.setAlignment(QtCore.Qt.AlignCenter)
        self.qlbl_cur_date_time = QtWid.QLabel("00-00-0000    00:00:00")
        self.qlbl_cur_date_time.setAlignment(QtCore.Qt.AlignCenter)
        self.qpbt_record = create_Toggle_button(
            "Click to start recording to file"
        )
        # fmt: off
        self.qpbt_record.clicked.connect(lambda state: log.record(state)) # pylint: disable=unnecessary-lambda
        # fmt: on

        vbox_middle = QtWid.QVBoxLayout()
        vbox_middle.addWidget(self.qlbl_title)
        vbox_middle.addWidget(self.qlbl_cur_date_time)
        vbox_middle.addWidget(self.qpbt_record)

        # Right box
        self.qpbt_exit = QtWid.QPushButton("Exit")
        self.qpbt_exit.clicked.connect(self.close)
        self.qpbt_exit.setMinimumHeight(30)
        self.qlbl_recording_time = QtWid.QLabel(alignment=QtCore.Qt.AlignRight)

        vbox_right = QtWid.QVBoxLayout()
        vbox_right.addWidget(self.qpbt_exit, stretch=0)
        vbox_right.addStretch(1)
        vbox_right.addWidget(self.qlbl_recording_time, stretch=0)

        # Round up top frame
        hbox_top = QtWid.QHBoxLayout()
        hbox_top.addLayout(vbox_left, stretch=0)
        hbox_top.addStretch(1)
        hbox_top.addLayout(vbox_middle, stretch=0)
        hbox_top.addStretch(1)
        hbox_top.addLayout(vbox_right, stretch=0)

        # -------------------------
        #   Bottom frame
        # -------------------------

        #  Charts
        # -------------------------

        self.gw = pg.GraphicsLayoutWidget()

        # Plot: Julabo temperatures
        p = {"color": "#EEE", "font-size": "10pt"}
        self.pi_julabo = self.gw.addPlot(row=0, col=0)
        self.pi_julabo.setLabel("left", text="temperature (°C)", **p)

        # Plot: Arduino temperatures
        self.pi_temp = self.gw.addPlot(row=1, col=0)
        self.pi_temp.setLabel("left", text="temperature (°C)", **p)

        # Plot: Arduino humidity
        self.pi_humi = self.gw.addPlot(row=2, col=0)
        self.pi_humi.setLabel("left", text="humidity (%)", **p)

        # Plot: Arduino pressure
        self.pi_pres = self.gw.addPlot(row=3, col=0)
        self.pi_pres.setLabel("left", text="pressure (mbar)", **p)

        self.plots = [self.pi_julabo, self.pi_temp, self.pi_humi, self.pi_pres]
        for plot in self.plots:
            plot.setClipToView(True)
            plot.showGrid(x=1, y=1)
            plot.setLabel("bottom", text="history (s)", **p)
            plot.setMenuEnabled(True)
            plot.enableAutoRange(axis=pg.ViewBox.XAxis, enable=False)
            plot.enableAutoRange(axis=pg.ViewBox.YAxis, enable=True)
            plot.setAutoVisible(y=True)
            plot.setRange(xRange=[-CHART_HISTORY_TIME, 0])

        # Curves
        capacity = round(CHART_HISTORY_TIME * 1e3 / DAQ_INTERVAL_MS)
        PEN_01 = pg.mkPen(color=[255, 255, 0], width=3)
        PEN_02 = pg.mkPen(color=[252, 15, 192], width=3)
        PEN_03 = pg.mkPen(color=[0, 255, 255], width=3)
        PEN_04 = pg.mkPen(color=[255, 255, 255], width=3)
        PEN_05 = pg.mkPen(color=[255, 127, 39], width=3)
        PEN_06 = pg.mkPen(color=[0, 255, 0], width=3)

        self.tscurve_julabo_setp = HistoryChartCurve(
            capacity=capacity,
            linked_curve=self.pi_julabo.plot(pen=PEN_05, name="Julabo setp."),
        )

        self.tscurve_julabo_bath = HistoryChartCurve(
            capacity=capacity,
            linked_curve=self.pi_julabo.plot(pen=PEN_06, name="Julabo bath"),
        )

        self.tscurve_ds_temp = HistoryChartCurve(
            capacity=capacity,
            linked_curve=self.pi_temp.plot(pen=PEN_01, name="DS temp."),
        )
        self.tscurve_bme_temp = HistoryChartCurve(
            capacity=capacity,
            linked_curve=self.pi_temp.plot(pen=PEN_02, name="BME temp."),
        )
        self.tscurve_bme_humi = HistoryChartCurve(
            capacity=capacity,
            linked_curve=self.pi_humi.plot(pen=PEN_03, name="BME humi."),
        )
        self.tscurve_bme_pres = HistoryChartCurve(
            capacity=capacity,
            linked_curve=self.pi_pres.plot(pen=PEN_04, name="BME pres."),
        )

        self.tscurves = [
            self.tscurve_julabo_setp,
            self.tscurve_julabo_bath,
            self.tscurve_ds_temp,
            self.tscurve_bme_temp,
            self.tscurve_bme_humi,
            self.tscurve_bme_pres,
        ]

        #  Group `Readings`
        # -------------------------

        legend = LegendSelect(
            linked_curves=self.tscurves, hide_toggle_button=True
        )

        p = {
            "readOnly": True,
            "alignment": QtCore.Qt.AlignRight,
            "maximumWidth": 54,
        }
        self.qlin_ds_temp = QtWid.QLineEdit(**p)
        self.qlin_bme_temp = QtWid.QLineEdit(**p)
        self.qlin_bme_humi = QtWid.QLineEdit(**p)
        self.qlin_bme_pres = QtWid.QLineEdit(**p)

        # fmt: off
        legend.grid.setHorizontalSpacing(6)
        legend.grid.addWidget(self.qlin_ds_temp       , 2, 2)
        legend.grid.addWidget(QtWid.QLabel("± 0.5 °C"), 2, 3)
        legend.grid.addWidget(self.qlin_bme_temp      , 3, 2)
        legend.grid.addWidget(QtWid.QLabel("± 0.5 °C"), 3, 3)
        legend.grid.addWidget(self.qlin_bme_humi      , 4, 2)
        legend.grid.addWidget(QtWid.QLabel("± 3 %")   , 4, 3)
        legend.grid.addWidget(self.qlin_bme_pres      , 5, 2)
        legend.grid.addWidget(QtWid.QLabel("± 1 mbar"), 5, 3)
        # fmt: on

        qgrp_readings = QtWid.QGroupBox("Readings")
        qgrp_readings.setLayout(legend.grid)

        #  Group 'Log comments'
        # -------------------------

        self.qtxt_comments = QtWid.QTextEdit()
        grid = QtWid.QGridLayout()
        grid.addWidget(self.qtxt_comments, 0, 0)

        qgrp_comments = QtWid.QGroupBox("Log comments")
        qgrp_comments.setLayout(grid)

        #  Group 'Charts'
        # -------------------------

        self.plot_manager = PlotManager(parent=self)
        self.plot_manager.add_autorange_buttons(linked_plots=self.plots)
        self.plot_manager.add_preset_buttons(
            linked_plots=self.plots,
            linked_curves=self.tscurves,
            presets=[
                {
                    "button_label": "01:00",
                    "x_axis_label": "history (sec)",
                    "x_axis_divisor": 1,
                    "x_axis_range": (-60, 0),
                },
                {
                    "button_label": "10:00",
                    "x_axis_label": "history (min)",
                    "x_axis_divisor": 60,
                    "x_axis_range": (-10, 0),
                },
                {
                    "button_label": "30:00",
                    "x_axis_label": "history (min)",
                    "x_axis_divisor": 60,
                    "x_axis_range": (-30, 0),
                },
                {
                    "button_label": "60:00",
                    "x_axis_label": "history (min)",
                    "x_axis_divisor": 60,
                    "x_axis_range": (-60, 0),
                },
                {
                    "button_label": "120:00",
                    "x_axis_label": "history (min)",
                    "x_axis_divisor": 60,
                    "x_axis_range": (-120, 0),
                },
            ],
        )
        self.plot_manager.add_clear_button(linked_curves=self.tscurves)
        self.plot_manager.perform_preset(1)

        qgrp_chart = QtWid.QGroupBox("Charts")
        qgrp_chart.setLayout(self.plot_manager.grid)

        vbox = QtWid.QVBoxLayout()
        vbox.addWidget(qgrp_readings)
        vbox.addWidget(qgrp_comments)
        vbox.addWidget(qgrp_chart, alignment=QtCore.Qt.AlignLeft)
        vbox.addStretch()

        # Round up bottom frame
        hbox_bot = QtWid.QHBoxLayout()
        hbox_bot.addWidget(self.gw, 1)
        hbox_bot.addLayout(vbox, 0)
        hbox_bot.addWidget(
            qdev_julabo.grpb, alignment=QtCore.Qt.AlignTop, stretch=0
        )

        # -------------------------
        #   Round up full window
        # -------------------------

        vbox = QtWid.QVBoxLayout(self)
        vbox.addLayout(hbox_top, stretch=0)
        vbox.addSpacerItem(QtWid.QSpacerItem(0, 10))
        vbox.addLayout(hbox_bot, stretch=1)

    # --------------------------------------------------------------------------
    #   Handle controls
    # --------------------------------------------------------------------------

    @QtCore.pyqtSlot()
    def update_GUI(self):
        str_cur_date, str_cur_time, _ = get_current_date_time()
        self.qlbl_cur_date_time.setText(
            "%s    %s" % (str_cur_date, str_cur_time)
        )
        self.qlbl_update_counter.setText("%i" % qdev_ard.update_counter_DAQ)
        self.qlbl_DAQ_rate.setText(
            "DAQ: %.1f Hz" % qdev_ard.obtained_DAQ_rate_Hz
        )
        if log.is_recording():
            self.qlbl_recording_time.setText(log.pretty_elapsed())

        self.qlin_ds_temp.setText("%.1f" % state.ds_temp)
        self.qlin_bme_temp.setText("%.1f" % state.bme_temp)
        self.qlin_bme_humi.setText("%.1f" % state.bme_humi)
        self.qlin_bme_pres.setText("%.1f" % state.bme_pres)

    @QtCore.pyqtSlot()
    def update_chart(self):
        if DEBUG:
            tprint("update_chart")

        for tscurve in self.tscurves:
            tscurve.update()


# ------------------------------------------------------------------------------
#   Program termination routines
# ------------------------------------------------------------------------------


def stop_running():
    app.processEvents()
    qdev_ard.quit()
    qdev_julabo.quit()
    log.close()

    print("Stopping timers................ ", end="")
    timer_GUI.stop()
    timer_charts.stop()
    print("done.")


@QtCore.pyqtSlot()
def notify_connection_lost():
    stop_running()

    window.qlbl_title.setText("! ! !    LOST CONNECTION    ! ! !")
    str_cur_date, str_cur_time, _ = get_current_date_time()
    str_msg = "%s %s\nLost connection to Arduino." % (
        str_cur_date,
        str_cur_time,
    )
    print("\nCRITICAL ERROR @ %s" % str_msg)
    reply = QtWid.QMessageBox.warning(
        window, "CRITICAL ERROR", str_msg, QtWid.QMessageBox.Ok
    )

    if reply == QtWid.QMessageBox.Ok:
        pass  # Leave the GUI open for read-only inspection by the user


@QtCore.pyqtSlot()
def about_to_quit():
    print("\nAbout to quit")
    stop_running()
    ard.close()


# ------------------------------------------------------------------------------
#   Your Arduino update function
# ------------------------------------------------------------------------------


def DAQ_function():
    # Date-time keeping
    str_cur_date, str_cur_time, str_cur_datetime = get_current_date_time()

    # Query the Arduino for its state
    success, tmp_state = ard.query_ascii_values("?", delimiter="\t")
    if not (success):
        dprint(
            "'%s' reports IOError @ %s %s"
            % (ard.name, str_cur_date, str_cur_time)
        )
        return False

    # Parse readings into separate state variables
    try:
        (
            state.time,
            state.ds_temp,
            state.bme_temp,
            state.bme_humi,
            state.bme_pres,
        ) = tmp_state
        state.time /= 1000  # Arduino time, [msec] to [s]
        state.bme_pres /= 100  # [Pa] to [mbar]
    except Exception as err:
        pft(err, 3)
        dprint(
            "'%s' reports IOError @ %s %s"
            % (ard.name, str_cur_date, str_cur_time)
        )
        return False

    # Catch very intermittent DS18B20 sensor errors
    if state.ds_temp <= -127.0:
        state.ds_temp = np.nan

    # We will use PC time instead
    state.time = time.perf_counter()

    # Add readings to chart histories
    window.tscurve_julabo_setp.appendData(state.time, julabo.state.setpoint)
    window.tscurve_julabo_bath.appendData(state.time, julabo.state.bath_temp)
    window.tscurve_ds_temp.appendData(state.time, state.ds_temp)
    window.tscurve_bme_temp.appendData(state.time, state.bme_temp)
    window.tscurve_bme_humi.appendData(state.time, state.bme_humi)
    window.tscurve_bme_pres.appendData(state.time, state.bme_pres)

    # Logging to file
    log.update(filepath=str_cur_datetime + ".txt", mode="w")

    # Return success
    return True


def write_header_to_log():
    log.write("[HEADER]\n")
    log.write(window.qtxt_comments.toPlainText())
    log.write("\n\n[DATA]\n")
    log.write("[s]\t[±0.5 °C]\t[±0.5 °C]\t[±3 pct]\t[±1 mbar]\t[°C]\t[°C]\n")
    log.write(
        "time\tDS_temp\tBME_temp\tBME_humi\tBME_pres\tJulabo_setp\tJulabo_bath\n"
    )


def write_data_to_log():
    log.write(
        "%.1f\t%.1f\t%.1f\t%.1f\t%.1f\t%.2f\t%.2f\n"
        % (
            log.elapsed(),
            state.ds_temp,
            state.bme_temp,
            state.bme_humi,
            state.bme_pres,
            julabo.state.setpoint,
            julabo.state.bath_temp,
        )
    )


# ------------------------------------------------------------------------------
#   Main
# ------------------------------------------------------------------------------

if __name__ == "__main__":
    # Set priority of this process to maximum in the operating system
    print("PID: %s\n" % os.getpid())
    try:
        proc = psutil.Process(os.getpid())
        if os.name == "nt":
            proc.nice(psutil.REALTIME_PRIORITY_CLASS)  # Windows
        else:
            proc.nice(-20)  # Other
    except:
        print("Warning: Could not set process to maximum priority.\n")

    # --------------------------------------------------------------------------
    #   Connect to devices
    # --------------------------------------------------------------------------

    # Arduino
    ard = Arduino(name="Ard", connect_to_specific_ID="BME280 & DS18B20 logger")
    ard.serial_settings["baudrate"] = 115200
    ard.auto_connect(filepath_last_known_port="config/port_Arduino.txt")

    if not (ard.is_alive):
        print("\nCheck connection and try resetting the Arduino.")
        print("Exiting...\n")
        sys.exit(0)

    # Julabo
    julabo = Julabo_circulator(name="Julabo")
    if julabo.auto_connect(filepath_last_known_port="config/port_Julabo.txt"):
        julabo.begin()

    # --------------------------------------------------------------------------
    #   Create application
    # --------------------------------------------------------------------------
    QtCore.QThread.currentThread().setObjectName("MAIN")  # For DEBUG info

    app = QtWid.QApplication(sys.argv)
    app.setFont(QtGui.QFont("Arial", 9))
    app.aboutToQuit.connect(about_to_quit)

    # --------------------------------------------------------------------------
    #   Set up multithreaded communication with the devices
    # --------------------------------------------------------------------------

    # Arduino
    qdev_ard = QDeviceIO(ard)
    qdev_ard.create_worker_DAQ(
        DAQ_function=DAQ_function,
        DAQ_interval_ms=DAQ_INTERVAL_MS,
        critical_not_alive_count=3,
        debug=DEBUG,
    )

    # Julabo
    qdev_julabo = Julabo_circulator_qdev(
        dev=julabo, DAQ_interval_ms=DAQ_INTERVAL_MS, debug=DEBUG
    )

    # --------------------------------------------------------------------------
    #   Create GUI
    # --------------------------------------------------------------------------

    window = MainWindow()

    # Connect signals
    qdev_ard.signal_DAQ_updated.connect(window.update_GUI)
    qdev_ard.signal_connection_lost.connect(notify_connection_lost)

    # --------------------------------------------------------------------------
    #   File logger
    # --------------------------------------------------------------------------

    log = FileLogger(
        write_header_function=write_header_to_log,
        write_data_function=write_data_to_log,
    )
    log.signal_recording_started.connect(
        lambda filepath: window.qpbt_record.setText(
            "Recording to file: %s" % filepath
        )
    )
    log.signal_recording_stopped.connect(
        lambda: window.qpbt_record.setText("Click to start recording to file")
    )

    # --------------------------------------------------------------------------
    #   Timers
    # --------------------------------------------------------------------------

    timer_GUI = QtCore.QTimer()
    timer_GUI.timeout.connect(window.update_GUI)
    timer_GUI.start(100)

    timer_charts = QtCore.QTimer()
    timer_charts.timeout.connect(window.update_chart)
    timer_charts.start(CHART_INTERVAL_MS)

    # --------------------------------------------------------------------------
    #   Start the main GUI event loop
    # --------------------------------------------------------------------------

    qdev_ard.start()
    qdev_julabo.start()

    window.show()
    sys.exit(app.exec_())
